libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+    add_segment_decr_jerk(tc, -Jc);
+    add_segment_incr_jerk(tc, Jc);
+    add_segment_decr_jerk(tc, Jc);
+
+    add_segment_incr_jerk(tc, Jc);
+    add_segment_decr_jerk(tc, Jc);
+    add_segment_incr_jerk(tc, -Jc);
+    add_segment_decr_jerk(tc, -Jc);
+
+    float Tcv = (Pm - 2.0 * (segment[num_segs - 1].start_pos - Pp)) / segment[num_segs - 1].start_vel;
+    add_segment_const_jerk(Tcv, 0.0);
+
+    add_segment_incr_jerk(tc, -Jc);
+    add_segment_decr_jerk(tc, -Jc);
+    add_segment_incr_jerk(tc, Jc);
+    add_segment_decr_jerk(tc, Jc);
+}
+
+// calculate duration of time segments for basic acceleration and deceleration curve from and to stationary.
+void scurves::cal_posfast(float tj, float Jp, float Ap, float Vp, float Pp, float &Jp_out, float &t2_out, float &t4_out, float &t6_out) const
+{
+    if ((Vp < Jp * (tj * tj) * 2.0) || (Pp < Jp * (tj * tj * tj) * 4.0)) {
+// solution = 0 - t6 t4 t2 = 0 0 0
+        t4_out = MIN(Vp / (2.0 * Jp * tj * tj), Pp / (4.0 * Jp * tj * tj * tj));
+        t2_out = 0;
+        t4_out = 0;
+        t6_out = 0;
+    } else if (Ap < Jp * tj) {
+// solution = 2 - t6 t4 t2 = 0 1 0
+        Jp = Ap / tj;
+        t2_out = 0;
+        t4_out = MIN((Vp - Ap * tj * 2.0) / Ap, -3.0 * tj + safe_sqrt((Pp * 2.0) / Ap + tj * tj));
+        t6_out = 0;
+    } else {
+        if ((Vp < Ap * tj + (Ap * Ap) / Jp) || (Pp < Ap * 1.0 / (Jp * Jp) * powf(Ap + Jp * tj, 2.0))) {
+// solution = 5 - t6 t4 t2 = 1 0 1
+            Ap = MIN(Ap, MIN(-Jp * tj * (1.0 / 2.0) + safe_sqrt(Jp * Vp * 4.0 + (Jp * Jp) * (tj * tj)) * (1.0 / 2.0), powf(Jp * tj - powf((Jp * Jp) * Pp * (1.0 / 2.0) + safe_sqrt((Jp * Jp * Jp * Jp) * (Pp * Pp) * (1.0 / 4.0) + (Jp * Jp * Jp * Jp * Jp) * Pp * (tj * tj * tj) * (1.0 / 2.7E1)) + (Jp * Jp * Jp) * (tj * tj * tj) * (1.0 / 2.7E1), 1.0 / 3.0) * 3.0, 2.0) * 1.0 / powf((Jp * Jp) * Pp * (1.0 / 2.0) + safe_sqrt((Jp * Jp * Jp * Jp) * (Pp * Pp) * (1.0 / 4.0) + (Jp * Jp * Jp * Jp * Jp) * Pp * (tj * tj * tj) * (1.0 / 2.7E1)) + (Jp * Jp * Jp) * (tj * tj * tj) * (1.0 / 2.7E1), 1.0 / 3.0) * (1.0 / 9.0)));
+            t2_out = Ap / Jp - tj;
+            t4_out = 0;
+            t6_out = t2_out;
+        } else {
+// solution = 7 - t6 t4 t2 = 1 1 1
+            t2_out = Ap / Jp - tj;
+            t4_out = MIN(Vp / Ap - (Ap + Jp * tj) / Jp, (Ap * (-3.0 / 2.0) - Jp * tj * (3.0 / 2.0)) / Jp + (safe_sqrt(Ap * Ap * Ap * Ap + (Ap * Ap) * (Jp * Jp) * (tj * tj) + Ap * (Jp * Jp) * Pp * 8.0 + (Ap * Ap * Ap) * Jp * tj * 2.0) * (1.0 / 2.0)) / (Ap * Jp));
+            t6_out = t2_out;
+        }
+    }
+    Jp_out = Jp;
+}
+
+// calculate the jerk, acceleration, velocity and position at time t
+bool scurves::update(float time, float &Jt_out, float &At_out, float &Vt_out, float &Pt_out)
+{
+    jtype_t Jtype;
+    int8_t pnt = num_segs;
+    float tj;
+    float Jp, T0, A0, V0, P0;
+
+    for (uint8_t i = 0; i < num_segs; i++) {
+        if (time < segment[num_segs - 1 - i].start_time) {
+            pnt = num_segs - 1 - i;
```

Comment:
Makes me nervous; if the caller screws up pnt can be -1 here and that's bad.

Reply:
Sorry, I don't understand this one.

Reply:
pnt changed to be uint8_t.

because pnt and num_segs can't be less than zero I think the contents of the for-loop are safe.  "i < num_segs" so "num_segs" must be at least "1" so , "segment[num_segs - 1 - i]" will be "segment[0]" which is OK.